File: keypress.js
var keypress = require('keypress');

// make `process.stdin` begin emitting "keypress" events
keypress(process.stdin);

function handle_keypress(ch, key) {
    if (key && key.ctrl && key.name == 'c') {
        process.stdin.pause();
        process.exit();
    } else if (key && key.name == 'return') {
        process.stdout.write('\r\n');
    } else if (key && key.name == 'backspace') {
        process.stdout.write('\033[1D \033[1D');
    } else {
        if (typeof ch != 'undefined') {
            process.stdout.write(ch);
        }
    }
}

// listen for the "keypress" event
process.stdin.on('keypress', function (ch, key) {
//  console.log('got "keypress"', key);

    handle_keypress(ch, key);
    msghub.publish('keypress', {ch: ch, key: key}, true);
});

process.stdin.setRawMode(true);
process.stdin.resume();


MsgHub = require('pubsub-client.js')
//MsgHub = require('/root/pubsub/pubsub-client.js')
//MsgHub = require('./pubsub/pubsub-client.js')
//msghub = new MsgHub('http://pubsub.msghub.io:58080')
msghub = new MsgHub('https://pubsub.msghub.io')
//msghub = new MsgHub('http://198.199.97.15:12345')

function cb(ch, msg) {
    handle_keypress(msg.ch, msg.key);
//    console.log('receive msg', msg, 'from channel', ch, (new Date()).toISOString());
}

msghub.subscribe('keypress', cb);
//msghub.publish('log', 'hello!!', false);

File: keypress.py
import termios, fcntl, sys, os
import json

#sys.path.append("./pubsub")
from pubsub_client import MsgHub


#msghub = MsgHub("pubsub.msghub.io", 443, secure=True)
msghub = MsgHub("https://pubsub.msghub.io", 443)
#msghub = MsgHub("pubsub.msghub.io", 12345)

def cb(ch, msg):
    handle_keypress(msg['ch'])

msghub.subscribe("keypress", cb)


def handle_keypress(ch):
    if ch == '\x7f':
        sys.stdout.write('\033[1D \033[1D')
    elif ch == '\x1b':
        pass
    else:
        sys.stdout.write(ch)

fd = sys.stdin.fileno()
oldterm = termios.tcgetattr(fd)
newattr = termios.tcgetattr(fd)
newattr[3] = newattr[3] & ~termios.ICANON & ~termios.ECHO
termios.tcsetattr(fd, termios.TCSANOW, newattr)

oldflags = fcntl.fcntl(fd, fcntl.F_GETFL)
fcntl.fcntl(fd, fcntl.F_SETFL, oldflags | os.O_NONBLOCK)

try:
    while True:
        try:
            c = sys.stdin.read(1)
            msghub.publish("keypress", {"ch": c, "key": None}, False)
#            msghub.publish("keypress", c, True)
            handle_keypress(c)
#            print "Got character", repr(c)
        except IOError: pass
finally:
    termios.tcsetattr(fd, termios.TCSAFLUSH, oldterm)
    fcntl.fcntl(fd, fcntl.F_SETFL, oldflags)
